### PR TITLE
[System] Check SocketOptionName.ReuseAddress support.

### PR DESCRIPTION
**Pull request by esdrubal**
Operating systems that use linux kernels before 3.9 do not support
binding multiple socket to the same address and port.

In those systems even if it was possible to set
SocketOptionName.ReuseAddress an exception was thrown on the second
bind.

These changes introduce a SocketException that is thrown when
SocketOptionName.ReuseAddress is set on a system that will fail to bind
multiple times to the same address.

### CODE REVIEW

File: mcs/class/System/Test/System.Net.Sockets/SocketTest.cs [hunk 1]
```diff
@@ -3464,32 +3464,32 @@ public void ReceiveRemoteClosed ()
...
-			ss.Bind (new IPEndPoint (IPAddress.Any, 12345));
+			using (Socket s = new Socket (AddressFamily.InterNetwork,
+						SocketType.Dgram, ProtocolType.Udp))
+			using (Socket ss = new Socket (AddressFamily.InterNetwork,
+						SocketType.Dgram, ProtocolType.Udp)) {
+				var supportsReuseAddress = true;
+				try {
+					s.SetSocketOption (SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+				} catch (SocketException e) {
+					// Exception is thrown when ReuseAddress is not supported
```
Comment (ludovic-henry) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: you should check here that e.NativeErrorCode == (int) SocketError.OperationNotSupported

File: mcs/class/System/Test/System.Net.Sockets/SocketTest.cs [hunk 2]
```diff
@@ -3500,12 +3500,19 @@ public void TcpDoubleBind ()
 						SocketType.Stream, ProtocolType.Tcp))
 			using (Socket ss = new Socket (AddressFamily.InterNetwork,
 						SocketType.Stream, ProtocolType.Tcp)) {
-				s.SetSocketOption (SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+				var supportsReuseAddress = true;
+				try {
+					s.SetSocketOption (SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+				} catch (SocketException e) {
+					// Exception is thrown when ReuseAddress is not supported
+					supportsReuseAddress = false;
```
Comment (ludovic-henry) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: same as above, you should check e.NativeErrorCode

**Comment by ludovic-henry** [comment]
Otherwise LGTM, merge whenever you want. Thank you!

**Comment by lewurm** [comment]
is there still interest to merge this or should it be closed?

**Comment by akoeplinger** [comment]
This broke a Socket test on Linux:

```
MESSAGE:
TtlChangeoverflow #3
+++++++++++++++++++
STACK TRACE:
at MonoTests.System.Net.Sockets.SocketTest.TtlChangeOverflow () [0x00039] in /mnt/jenkins/workspace/test-mono-mainline-linux/label/ubuntu-1404-amd64/mcs/class/System/Test/System.Net.Sockets/SocketTest.cs:1013
at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /mnt/jenkins/workspace/test-mono-mainline-linux/label/ubuntu-1404-amd64/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
```

**Pull request merged**